File: 0020-0029/0021.rb
# Wrong Answerになるので一旦放置
def parallels?(x1, y1, x2, y2, x3, y3, x4, y4)
  a1 = (y2-y1)/(x2-x1) rescue Float::INFINITY
  a2 = (y4-y3)/(x4-x3) rescue Float::INFINITY
  return a1 == a2
end

n = gets.to_i
n.times do
  x1, y1, x2, y2, x3, y3, x4, y4 = gets.split(' ').map(&:to_f)
  if (y1==y2 && y3==y4) or parallels?(x1, y1, x2, y2, x3, y3, x4, y4)
    puts 'YES'
  else
    puts 'NO'
  end
end

File: python/0021.py
# Wrong Answerになるので一旦放置
n = int(input()) + 1
for val in range(1,n):
  x = map(float, raw_input().split(' '))
  if x[1]-x[3] == 0 and x[5]-x[7] == 0:
    print 'YES'
  elif x[1]-x[3] != 0 and x[5]-x[7] == 0:
    print 'NO'
  elif x[1]-x[3] == 0 and x[5]-x[7] != 0:
    print 'NO'
  else:
    if (x[0]-x[2])/(x[1]-x[3]) == (x[4]-x[6])/(x[5]-x[7]):
      print 'YES'
    else:
      print 'NO'